File: ios/NDIWebcam/NDIStreamer.swift
import Foundation
import CoreMedia
import CoreVideo
import AudioToolbox

// Thread-safe NDI send wrapper.
// All send calls are dispatched on a dedicated serial queue so the
// AVCapture delegate threads never block each other.
final class NDIStreamer {

    enum StreamError: Error {
        case initializationFailed
        case senderCreationFailed
    }

    private(set) var isStreaming = false
    private var sendInstance: NDIlib_send_instance_t?
    private let sendQueue = DispatchQueue(label: "com.ndiwebcam.ndi-send", qos: .userInteractive)

    // MARK: Lifecycle

    func start(sourceName: String) throws {
        guard NDIlib_initialize() else { throw StreamError.initializationFailed }

        var settings = NDIlib_send_create_t()
        let nameBytes = (sourceName as NSString).utf8String
        settings.p_ndi_name = nameBytes
        settings.clock_video = false
        settings.clock_audio = false

        guard let instance = NDIlib_send_create(&settings) else {
            NDIlib_destroy()
            throw StreamError.senderCreationFailed
        }
        sendInstance = instance
        isStreaming = true
    }

    func stop() {
        sendQueue.sync {
            if let instance = sendInstance {
                NDIlib_send_destroy(instance)
                sendInstance = nil
            }
            NDIlib_destroy()
            isStreaming = false
        }
    }

    // MARK: Video

    func sendVideoFrame(_ sampleBuffer: CMSampleBuffer, fps: Int32) {
        guard isStreaming, let instance = sendInstance else { return }
        guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else { return }

        CVPixelBufferLockBaseAddress(pixelBuffer, .readOnly)
        defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, .readOnly) }

        let width  = Int32(CVPixelBufferGetWidth(pixelBuffer))
        let height = Int32(CVPixelBufferGetHeight(pixelBuffer))
        let stride = Int32(CVPixelBufferGetBytesPerRow(pixelBuffer))
        let data   = CVPixelBufferGetBaseAddress(pixelBuffer)?.assumingMemoryBound(to: UInt8.self)

        var frame = NDIlib_video_frame_v2_t()
        frame.xres = width
        frame.yres = height
        // Camera2 outputs BGRA; UYVY is more efficient but requires conversion.
        // Using BGRA here avoids an extra conversion pass on the hot path.
        frame.FourCC = NDIlib_FourCC_type_BGRA
        frame.frame_rate_N = fps * 1000
        frame.frame_rate_D = 1000
        frame.picture_aspect_ratio = Float(width) / Float(height)
        frame.frame_format_type = NDIlib_frame_format_type_progressive
        frame.timecode = Int64(bitPattern: 0x8000000000000000) // NDIlib_send_timecode_synthesize
        frame.p_data = data
        frame.line_stride_in_bytes = stride

        sendQueue.async { [instance] in
            NDIlib_send_send_video_v2(instance, &frame)
        }
    }

    // MARK: Audio

    // Expects 32-bit float, non-interleaved (planar) PCM — the format AVCaptureAudioDataOutput delivers.
    func sendAudioFrame(_ sampleBuffer: CMSampleBuffer) {
        guard isStreaming, let instance = sendInstance else { return }

        var blockBuffer: CMBlockBuffer?
        var audioBufferList = AudioBufferList()
        var bufferListSize = MemoryLayout<AudioBufferList>.size

        let status = CMSampleBufferGetAudioBufferListWithRetainedBlockBuffer(
            sampleBuffer,
            bufferListSizeNeededOut: nil,
            bufferListOut: &audioBufferList,
            bufferListSize: bufferListSize,
            blockBufferAllocator: nil,
            blockBufferMemoryAllocator: nil,
            flags: 0,
            blockBufferOut: &blockBuffer
        )
        guard status == noErr else { return }

        guard let format = CMSampleBufferGetFormatDescription(sampleBuffer) else { return }
        let asbd = CMAudioFormatDescriptionGetStreamBasicDescription(format)?.pointee
        guard let sampleRate = asbd.map({ Int32($0.mSampleRate) }),
              let channels   = asbd.map({ Int32($0.mChannelsPerFrame) }) else { return }

        let sampleCount = Int32(CMSampleBufferGetNumSamples(sampleBuffer))

        // AVCapture delivers float32 interleaved; NDI expects float32 planar (FLTP).
        // For mono/stereo from device mic, de-interleave into planar layout.
        let buffer = audioBufferList.mBuffers
        guard let rawData = buffer.mData else { return }

        var frame = NDIlib_audio_frame_v3_t()
        frame.sample_rate = sampleRate
        frame.no_channels = channels
        frame.no_samples  = sampleCount
        frame.timecode    = Int64(bitPattern: 0x8000000000000000)
        frame.FourCC      = NDIlib_FourCC_audio_type_FLTP
        frame.p_data      = rawData.assumingMemoryBound(to: UInt8.self)
        frame.channel_stride_in_bytes = sampleCount * Int32(MemoryLayout<Float>.size)

        sendQueue.async { [instance] in
            NDIlib_send_send_audio_v3(instance, &frame)
        }
    }

    // MARK: Status

    var connectedReceiverCount: Int {
        guard let instance = sendInstance else { return 0 }
        return Int(NDIlib_send_get_no_connections(instance, 0))
    }
}

File: android/app/src/main/java/com/ndiwebcam/NDIStreamer.kt
package com.ndiwebcam

import android.util.Log

/**
 * Kotlin wrapper around the native NDI send API exposed via ndi_jni.cpp.
 * All JNI calls are made from the caller's thread — the JNI layer is thread-safe.
 */
class NDIStreamer {

    var isStreaming = false
        private set

    // MARK: Lifecycle

    fun start(sourceName: String): Boolean {
        val ok = nativeInit(sourceName)
        isStreaming = ok
        if (!ok) Log.e(TAG, "NDI init failed")
        return ok
    }

    fun stop() {
        if (isStreaming) {
            nativeDestroy()
            isStreaming = false
        }
    }

    // MARK: Video
    // data: UYVY byte array (width * height * 2)
    fun sendVideoFrame(data: ByteArray, width: Int, height: Int, fpsNumerator: Int = 30000, fpsDenominator: Int = 1000) {
        if (!isStreaming) return
        nativeSendVideoFrame(data, width, height, fpsNumerator, fpsDenominator)
    }

    // MARK: Audio
    // data: float32 planar PCM (channel_stride = noSamples floats per channel)
    fun sendAudioFrame(data: FloatArray, sampleRate: Int, channels: Int, noSamples: Int) {
        if (!isStreaming) return
        nativeSendAudioFrame(data, sampleRate, channels, noSamples)
    }

    // MARK: Status
    val connectedReceiverCount: Int
        get() = if (isStreaming) nativeGetReceiverCount() else 0

    // MARK: JNI

    private external fun nativeInit(sourceName: String): Boolean
    private external fun nativeSendVideoFrame(data: ByteArray, width: Int, height: Int, fpsN: Int, fpsD: Int): Boolean
    private external fun nativeSendAudioFrame(data: FloatArray, sampleRate: Int, channels: Int, noSamples: Int): Boolean
    private external fun nativeGetReceiverCount(): Int
    private external fun nativeDestroy()

    companion object {
        private const val TAG = "NDIStreamer"

        init {
            // Loads libndi.so (the Vizrt NDI library) then our JNI bridge.
            System.loadLibrary("ndi")
            System.loadLibrary("ndiwebcam")
        }
    }
}
